Adds subgraph volume-computation, support for directed graphs

fabratu opened the pull request:
This PR would add the following:

- Adds support for volume of directed graphs. Implementation following initial thoughts from https://github.com/networkit/networkit/pull/558#issuecomment-632704544
- Adds support for volume of subgraphs. Implementation following initial thoughts from https://github.com/networkit/networkit/pull/558#discussion_r430247248
- Adds unit-tests and python-interface https://github.com/networkit/networkit/pull/558#pullrequestreview-421021520

fabratu commented:
PR is ready for review (error in doc-build fixed in #590 )

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+    // Compute volume of subgraph
+    if (!nodes.empty() > 0 && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            outVolume += G.weightedDegree(*it, directed);
```
Counting self-loops twice for directed graphs seems wrong. The version for the whole graph counts them twice only for undirected graphs (as I would expect). I think this should be `!directed`. Or maybe not, as the variable is actually `true` if the graph is undirected?!

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+    // Compute volume of subgraph
+    if (!nodes.empty() && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            inVolume += G.weightedDegreeIn(*it, directed);
```
Counting self-loops twice for directed graphs seems wrong. The version for the whole graph counts them twice only for undirected graphs (as I would expect). I think this should be `!directed`. Or maybe not, as the variable is actually `true` if the graph is undirected?!

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+            G.forInNeighborsOf(*it, [&](node v) {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    inVolume -= static_cast<double>(G.weight(v, *it));
```
The edge weight should instead be taken as parameter in this lambda by defining it as `[&](node, node v, edgeweight w)`.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() && G.numberOfNodes() != nodes.size()) {
```
While the optimization for not counting the whole graph seems neat, I wonder if we should not check that the graph really contains all nodes that are in the set.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+
+    // Compute volume of subgraph
+    if (!nodes.empty() && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
```
I would suggest a quick check if the node actually exists here as the performance impact should be minimal (given that we iterate over all neighbors, anyway).

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    inVolume -= static_cast<double>(G.weight(v, *it));
+                }
```
For weighted graphs, this iterates over the neighbors twice: Once to collect the edge weights in `weightedDegree` and once here to remove some of them again. Why don't we directly only add those edge weights to neighbors that are in the set, i.e., remove the call to `G.weightedDegree` and add weights in the second loop when `neighborIt != nodes.end()`.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() > 0 && G.numberOfNodes() != nodes.size()) {
```
Same here: While the optimization for not counting the whole graph seems neat, I wonder if we should not check that the graph really contains all nodes that are in the set.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    outVolume -= static_cast<double>(G.weight(*it, v));
+                }
```
Same here: For weighted graphs, this iterates over the neighbors twice: Once to collect the edge weights in `weightedDegree` and once here to remove some of them again. Why don't we directly only add those edge weights to neighbors that are in the set, i.e., remove the call to `G.weightedDegree` and add weights in the second loop when `neighborIt != nodes.end()`.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+double inVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
```
Do you mean `bool directed = !G.isDirected()`? If yes, this is extremely confusing. Please rename the variable to `undirected`.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
-        return 2.0 * G.numberOfEdges();
+double outVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double outVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
```
Do you mean `bool directed = !G.isDirected()`? If yes, this is extremely confusing. Please rename the variable to `undirected`.

michitux reviewed (changes requested):
Why is the method `volume` removed without any deprecation? Further, the term _volume_ of a subset of nodes as I know it is usually the sum of the weights of the edges incident to the nodes, see, e.g., [Wikipedia on Conductance](https://en.wikipedia.org/wiki/Conductance_(graph)).

This is two times almost the same code. I would suggest to refactor this code to just one implementation.

fabratu commented:
Thanks for the thorough review, this helps a lot.

While most of it is clear, I would suggest leave two implementations. This split into out/in-functions is used in several places in nkt. However I changed the naming to align with other functions like `weightedDegree`. Also re-introduced the `volume`-function and split subgraph/graph-volume computation. Does it look better now?

angriman commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -201,14 +201,65 @@ double density(const Graph &G) noexcept {
+    count nodesNotInGraph = 0;
+
+    // Check for nodes in subset, which are not part of the graph
+    for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
```
What about `for (node u : nodes)`?

angriman replied:
Of course my comment does not hold anymore if we are going to use ranges for the input nodes.

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,36 @@ std::pair<node, node> size(const Graph &G) noexcept;
+ * @param nodes Subset of nodes.
+ * @return inVolume of the subgraph.
+ */
+double subgraphInVolume(const Graph &G, const std::unordered_set<node> &nodes);
```
For the input nodes, can we use ranges? So the user is not forced to create an `unordered_set`.

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,36 @@ std::pair<node, node> size(const Graph &G) noexcept;
+ * @param nodes Subset of nodes.
+ * @return Volume of the graph.
+ */
+double subgraphVolume(const Graph &G, const std::unordered_set<node> &nodes);
```
Same here.

michitux commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -201,14 +201,65 @@ double density(const Graph &G) noexcept {
+    } 
+
+    // Compute volume of subgraph if node-subset is non-empty
+    if (!nodes.empty() > 0 && (G.numberOfNodes() - nodesNotInGraph) != nodes.size()) {
```
The `> 0` in the first condition should be removed. The `nodesNotInGraph` should be subtracted from `nodes.size()`, not `G.numberOfNodes()`.

michitux commented on networkit/cpp/graph/test/GraphToolsGTest.cpp:
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
+        double baseG2 = weighted() ? static_cast<double>(G2.totalEdgeWeight())
+                                   : static_cast<double>(G2.numberOfEdges());
+        EXPECT_NEAR(outVolumeSub, mod * baseG2, 1e-7);
+        EXPECT_NEAR(inVolumeSub, mod * baseG2, 1e-7);
```
This does not match the new definition of volume and the test fails accordingly.

michitux reviewed (changes requested):
This looks much better. I agree with @angriman concerning the use of ranges, in particular as with the new implementation, no set membership needs to be tested anymore. I am not so sure about the names of the functions, though, as this is not the volume of a subgraph, but the volume of a set of nodes. Maybe just leaving them as `volume`/`inVolume` makes more sense?

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
  *
  * @param G The input graph.
  * @return Volume of the graph.
+ * 
```
This change is not needed

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
+ * of nodes.
+ *
+ * @param G The input graph.
+ * @param first Iterator for the starting node in the set.
```
Maybe it is clearer `@param first,last The range of nodes to insert` as they do in cppreference (e.g., https://en.cppreference.com/w/cpp/container/vector/insert).

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
+ * @param G The input graph.
+ * @param first Iterator for the starting node in the set.
+ * @param last Iterator for the last node in the set.
+ * @return volume of the graph.
```
```suggestion
 * @return volume of the graph.
```
```suggestion
 * @return Volume of the graph.
```

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
+    double outVolume = 0.0;
+    bool countSelfLoopsTwice = !G.isDirected();
+    count nodesInGraph = 0;
+    InputIt actIt = first;
```
Probably it is cleaner to simply declare `actIt` in the for loops below. E.g., `for (auto actIt = first; actIt != last; ++actIt)`

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
+ * @param first Iterator for the starting node in the set.
+ * @param last Iterator for the last node in the set.
+ * @return inVolume of the subgraph.
+ */
```
Same comments as in `volume`, both code and documentation.

angriman commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -201,14 +201,17 @@ double density(const Graph &G) noexcept {
 double volume(const Graph &G) {
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+    auto volume = 0.0;
```
I don't think there is a need for `auto` here, simply use `double`.

angriman commented on networkit/cpp/graph/test/GraphToolsGTest.cpp:
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
+        double mod = directed() ? 1.0 : 2.0;
+        double baseG = weighted() ? static_cast<double>(G.totalEdgeWeight())
+                                  : static_cast<double>(G.numberOfEdges());
+        EXPECT_NEAR(volume, mod * baseG, 1e-7);
```
What about `EXPECT_DOUBLE_EQ`?

angriman commented on networkit/graphtools.pyx:
```diff
@@ -337,6 +339,46 @@ cdef class GraphTools:
+			The inVolume of the input graph.
+		"""
+		return subgraphInVolume(graph._this, nodes)
+
```
Could you please add some simple Python tests for these APIs?

fabratu commented:
Thanks for the additional input, took care of everything.

angriman commented on networkit/cpp/graph/test/GraphToolsGTest.cpp:
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
+        const auto volumeComplementSet = GraphTools::volume(G, lastOfSet, last);
+        const auto inVolumeComplementSet= GraphTools::inVolume(G, lastOfSet, last);
+
+        EXPECT_NEAR(volumeSet + volumeComplementSet, volume, 1e-7);
+        EXPECT_NEAR(inVolumeSet + inVolumeComplementSet, volume, 1e-7);
```
I guess you can use `EXPECT_DOUBLE_EQ` here as well right?

fabratu replied:
No, there are differences in the range of approx. 1e-5

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,79 @@ std::pair<node, node> size(const Graph &G) noexcept;
+            nodesInGraph++;
+    } 
+
+    // Compute volume of subgraph if the subset is really a subset
+    if (nodesInGraph != G.numberOfNodes()) {
+        for (auto actIt = first; actIt != last; ++actIt) {
+            outVolume += G.weightedDegree(*actIt, countSelfLoopsTwice);
+        }
+    // Otherwise compute volume of the complete graph
+    } else {
+        outVolume = volume(G);
+    }
```
Does it really make sense to distinguish these two cases? I think that you can simply compute the volume as
```
for (auto actIt = first; actIt != last; ++actIt) {
    outVolume += G.weightedDegree(*actIt, countSelfLoopsTwice);
}
```
and we can assume that the input range is already a set, we should not check that.

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,79 @@ std::pair<node, node> size(const Graph &G) noexcept;
+    } 
+
+    // Compute volume of subgraph if the subset is really a subset
+    if (nodesInGraph != G.numberOfNodes()) {
+        for (auto actIt = first; actIt != last; ++actIt) {
+            inVolume += G.weightedDegreeIn(*actIt, countSelfLoopsTwice);
+        }
+    // Otherwise compute volume of the complete graph
+    } else {
+        inVolume = volume(G);
+    }
```
Same thing here

angriman commented on networkit/cpp/graph/GraphTools.cpp:
```diff
@@ -201,14 +201,17 @@ double density(const Graph &G) noexcept {
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+    if (G.isWeighted()) {
+        volume = G.totalEdgeWeight();
     } else {
-        return 2.0 * G.totalEdgeWeight();
+        volume = G.numberOfEdges();
+    }
```
`G.totalEdgeWeight()` already handles the weighted/unweighted case, you can simply use `double volume = G.totalEdgeWeight()` and drop the `if`.

angriman commented on networkit/cpp/graph/test/GraphToolsGTest.cpp:
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
+ 
+        // Volume for subgraph G2
+        node seed = Aux::Random::integer(G.upperNodeIdBound());
+        auto sNodes = std::unordered_set<node>();
```
This is never used

angriman commented on networkit/cpp/graph/test/GraphToolsGTest.cpp:
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
+        node seed = Aux::Random::integer(G.upperNodeIdBound());
+        auto sNodes = std::unordered_set<node>();
+
+        Graph::NodeIterator first = Graph::NodeIterator(&G, 0);
+        Graph::NodeIterator lastOfSet = Graph::NodeIterator(&G, seed);
+        Graph::NodeIterator last = Graph::NodeIterator(&G, G.upperNodeIdBound());
+
+        const auto volumeSet = GraphTools::volume(G, first, lastOfSet);
```
This is not really needed, why don't you just use something like
```
const auto volumeSet = GraphTools::volume(G, G.nodeRange().begin(), G.nodeRange().begin() + seed);
```

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,55 @@ std::pair<node, node> size(const Graph &G) noexcept;
+    double inVolume = 0.0;
+
+    // Compute volume of subgraph
+    for (auto actIt = first; actIt != last; ++actIt) {
+        inVolume += G.weightedDegreeIn(*actIt, !G.isDirected());
+    }
```
Even simpler:
```
for (; first != last; ++first)
    inVolume += G. weightedDegreeIn(*first, !G.isDirected());
```

angriman commented on include/networkit/graph/GraphTools.hpp:
```diff
@@ -129,13 +129,55 @@ std::pair<node, node> size(const Graph &G) noexcept;
+    double outVolume = 0.0;
+
+    // Compute volume of subgraph
+    for (auto actIt = first; actIt != last; ++actIt) {
+        outVolume += G.weightedDegree(*actIt, !G.isDirected());
+    }
```
I wrote below a simpler version of this loop.

angriman commented on networkit/cpp/graph/test/GraphToolsGTest.cpp:
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
+
+        // Volume for either directed/undirected
+        double mod = directed() ? 1.0 : 2.0;
+        double baseG = weighted() ? static_cast<double>(G.totalEdgeWeight())
```
The type of `totalEdgeWeight` is `edgeweight` (double), you don't need to cast it to double.

angriman commented:
These changes need to be approved also by @michitux

michitux reviewed (dismissed):
There are still some minor issues with the Python interface, see individual comments. Apart from that I approve this.

fabratu commented on networkit/graphtools.pyx:
```diff
@@ -321,21 +323,47 @@ cdef class GraphTools:
+			The volume of the subgraph.
 		"""
-		return volume(graph._this)
+		if nodes.size() == 0:
```
Thanks, yeah this would be cleaner. Adopted the changes (also the other comment).

michitux commented on networkit/graphtools.pyx:
```diff
@@ -321,21 +323,61 @@ cdef class GraphTools:
+			return volume(graph._this)
+
+	@staticmethod
+	def inVolume(Graph graph, nodes = None):
```
If for `inVolume`  the nodes parameter is not optional, why is it declared with a default value of `None`, which just raises an exception below?

fabratu replied:
True. Thanks for all the input and being more careful than me.

michitux commented:
I see no further problems with this code. There is just one minor thing I would like to mention which can definitely wait for another round of improvements: Currently, the GIL is not released during the computation of the volume. This means that Python code in other threads cannot run in parallel. This would be possible though. What would be required is to first save the return value in a C++ variable, i.e., a variable declared, e.g.,  with `cdef double result`. Then the call to volume/inVolume could be wrapped in a `with nogil` block in which no Python objects are touched anymore and thus no GIL is required.
